Upstream

AniketSarkar123 opened the pull request:
1. Added a new mailbox driver for k3-low socs like AM62L as sec_proxy is not used.
2. Code abstracted for the two communicating layers

AniketSarkar123 committed:
plat-k3: drivers: Add support for TI mailbox driver

New devices like the AM62L will use a mailbox to communicate with the
security firmware. Add mailbox driver here to support the mailbox
hardware.

Signed-off-by: Aniket Sarkar <a-sarkar1@ti.com>

AniketSarkar123 committed:
plat-k3: drivers: Refactor sec_proxy driver

Since the underneath transport layer may or may not always be sec_proxy
it doesn't make sense to keep following the k3_sec_proxy_* convention
for the TI_SCI message transports.
Rename them to something more generic like ti_sci_transport_*.

Signed-off-by: Aniket Sarkar <a-sarkar1@ti.com>

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,192 @@
+	paddr_t pa;
+
+	if (!msg)
+		return -1;
```
Return error codes (TEE_SUCCESS or TEE_ERROR_*) instead of 0 or -1 in all places functions returning a TEE_Result

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,192 @@
+	uint8_t *rcv_va;
+
+	if (!msg)
+		return -1;
```
Same here. Change returns.

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,192 @@
+/*
```
Add a SPDX License Identifier

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,192 @@
+		io_read32((vaddr_t)mailbox_rx_base + TI_MAILBOX_MSG);
+		if (try_count == 0U) {
+			EMSG("Mailbox RX status polling timed out");
+			return -1;
```
Cannot return -1 if the return type is uint32_t

Pratham-T replied:
After changing return type to TEE_Result, return propet TEE_ERROR_* macro code.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -15,10 +15,11 @@
-#include "ti_sci_protocol.h"
+#include<ti_sci_transport.h>
+#define TI_SCI_MAX_MESSAGE_SIZE		56
+ static uint8_t message_seq=0;
```
White-space issues. Please run checkpatch to catch these easily.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -57,18 +58,23 @@ static int ti_sci_setup_xfer(uint16_t msg_type, uint32_t msg_flags,
-	    rx_message_size < sizeof(*hdr) ||
+	if (rx_message_size > TI_SCI_MAX_MESSAGE_SIZE ||
+	    tx_message_size > TI_SCI_MAX_MESSAGE_SIZE ||
+	     rx_message_size < sizeof(*hdr) ||
```
Extra whitespace here as well.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -57,18 +58,23 @@ static int ti_sci_setup_xfer(uint16_t msg_type, uint32_t msg_flags,
+	hdr->host = TI_SCI_HOST_ID;
+	hdr->flags = msg_flags;
+	if(rx_message_size!=0U){
+		hdr->flags |= TI_SCI_FLAG_REQ_ACK_ON_PROCESSED;
```
Lot of whitespace inconsistencies. Follow coding guideines:
[OPTEE Coding standards](https://optee.readthedocs.io/en/latest/general/coding_standards.html#coding-standards)
[Linux Kernel Coding Style](https://www.kernel.org/doc/html/latest/process/coding-style.html)

Run checkpatch script (see above links) to catch these.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -88,41 +94,43 @@ static int ti_sci_setup_xfer(uint16_t msg_type, uint32_t msg_flags,
-	message_sequence++;
-	txhdr->seq = message_sequence;
+	ret=ti_sci_clear_init();
+	if(ret){
```
Again, not following coding style guidelines.

Pratham-T commented on core/include/ti_sci_transport.h:
```diff
@@ -0,0 +1,68 @@
+/*
```
Add proper SPDX License Identifier

Pratham-T commented on core/include/ti_sci_transport.h:
```diff
@@ -0,0 +1,68 @@
+ * Return: 0 on success, or an error code on failure.
+ */
+
+uint32_t ti_sci_transport_send(const struct ti_sci_msg *msg);
```
Return type should be TEE_Result. Which is what is used in function definition in sec_proxy.c file.

Pratham-T commented on core/include/ti_sci_transport.h:
```diff
@@ -0,0 +1,68 @@
+ * Return: 0 on success, or an error code on failure.
+ */
+
+uint32_t ti_sci_transport_recv(const struct ti_sci_msg *msg);
```
Same here. Change return type. Use OPTEE GlobalPlatform types.

Pratham-T commented on core/include/ti_sci_transport.h:
```diff
@@ -0,0 +1,68 @@
+ * Return: 0 on success, or an error code on failure.
+ */
+
+uint32_t ti_sci_clear_init(void);
```
Same here. Change return type.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -527,6 +538,8 @@ int ti_sci_init(void)
 	struct ti_sci_msg_resp_version rev_info = { };
 	int ret = 0;
+	// generic_delay_timer_init();
+	// ti_init_scmi_server();
```
Remove unused commented code.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -57,18 +58,23 @@ static int ti_sci_setup_xfer(uint16_t msg_type, uint32_t msg_flags,
 	hdr->type = msg_type;
-	hdr->host = OPTEE_HOST_ID;
-	hdr->flags = msg_flags | TI_SCI_FLAG_REQ_ACK_ON_PROCESSED;
+	hdr->host = TI_SCI_HOST_ID;
```
TI_SCI_HOST_ID macro is from Arm Trusted Firmware. It is not defined anywhere in OPTEE nor is introduced by you. Moreover, the host ID to be used here will be OPTEE_HOST_ID only since the host here is OPTEE and not ATF. Revert this back.

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,192 @@
+#include <ti_sci_protocol.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
```
It is a generally followed convention that imports are ordered alphabetically

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,192 @@
+ * ti_sci_clear_init() - Clear the mailbox RX status and initialize the mailbox
+ */
+
+uint32_t ti_sci_clear_init(void)
```
Change return type to TEE_Result

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,192 @@
+	}
+
+	IMSG("Mailbox MMIO and SRAM regions mapped: rx=%p tx=%p tx_sram=%p rx_sram=%p",
+	      mailbox_rx_base, mailbox_tx_base, mailbox_tx_sram_va, mailbox_rx_sram_va);
```
Function arguments split in multiple lines, not aligned correctly here.

Pratham-T reviewed (changes requested):
Reviewed for general errors and mistakes, and coding style violations. Did not review mailbox implementation correctness.

Also, commit split is not logical. First commit should be refactoring existing code which also introduces ti_sci_transport.h. This header file needs to be included in the mailbox.c file which is currently not included. Introducing mailbox commit should hence be second after refactoring.

Please make all these corrections.

Pratham-T commented:
Commit order is still incorrect. [refactor(ti): abstract sec_proxy and mailbox drivers.](https://github.com/AniketSarkar123/optee_os/pull/1/commits/57539402cc543b1d2b8c93e0341578b92a1eee2b) adds non-existent mailbox.c in sub.mk, which is being created in [feat(ti): add support for TI mailbox driver](https://github.com/AniketSarkar123/optee_os/pull/1/commits/c849d7e714fe059022fa46b582228f8850cb6c57)

Also, drop period at the end of the commit message header.

AniketSarkar123 commented:
Ok correcting that

On Tue, Jul 8, 2025, 13:27 T Pratham ***@***.***> wrote:

> *Pratham-T* left a comment (AniketSarkar123/optee_os#1)
> <https://github.com/AniketSarkar123/optee_os/pull/1#issuecomment-3047785344>
>
> Commit order is still incorrect. refactor(ti): abstract sec_proxy and
> mailbox drivers.
> <https://github.com/AniketSarkar123/optee_os/pull/1/commits/57539402cc543b1d2b8c93e0341578b92a1eee2b>
> adds non-existent mailbox.c in sub.mk, which is being created in feat(ti):
> add support for TI mailbox driver
> <https://github.com/AniketSarkar123/optee_os/pull/1/commits/c849d7e714fe059022fa46b582228f8850cb6c57>
>
> Also, drop period at the end of the commit message header.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AniketSarkar123/optee_os/pull/1#issuecomment-3047785344>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A4XDCCJQ67TTRF6UOKCXBFL3HN2VVAVCNFSM6AAAAACA5VXBWOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANBXG44DKMZUGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ti-kamlesh commented on core/arch/arm/plat-k3/platform_config.h:
```diff
@@ -54,6 +54,13 @@
+#define MAILBOX_RX_START_REGION 0x70815000UL
+#define MAILBOX_MAX_MESSAGE_SIZE	56U
+#define MAILBOX_TX_REGION_SIZE   0x7081413FUL
 #define SEC_PROXY_TIMEOUT_US            1000000
```
align the hex numbers

ti-kamlesh commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,194 @@
+
+	try_count = io_read32((vaddr_t)mailbox_rx_base + TI_MAILBOX_MSG_STATUS);
+	while (io_read32((vaddr_t)mailbox_rx_base +
+	TI_MAILBOX_MSG_STATUS) != 0U) {
```
is ths split from earlier really required? if yes, aliign properly

ti-kamlesh commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,194 @@
+ * mailbox_init() - Initialize the mailbox MMIO and SRAM regions
+ */
+
+void mailbox_init(void)
```
Blank line after comments is this a standard ?

ti-kamlesh commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,194 @@
+	while (num_messages_pending == 0U) {
+		num_messages_pending = io_read32((vaddr_t)mailbox_rx_base +
+		TI_MAILBOX_MSG_STATUS);
+
```
alignment

Pratham-T replied:
This is still not resolved here. @AniketSarkar123

ti-kamlesh commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,194 @@
+		if (retry_count-- == 0U) {
+			EMSG("Mailbox RX status polling timed out");
+			return -1;
+		}
```
Missed replacing -1

ti-kamlesh commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,194 @@
+	}
+
+	return 0;
+}
```
Add proper return Macro here

AniketSarkar123 commented:
Ok seeing the issues

On Tue, Jul 15, 2025, 16:38 ti-kamlesh ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In core/arch/arm/plat-k3/drivers/mailbox.c
> <https://github.com/AniketSarkar123/optee_os/pull/1#discussion_r2207165029>
> :
>
> > +/*
> + * TI Mailbox MMIO and SRAM base addresses
> + * These are the base addresses for the mailbox MMIO and SRAM regions.
> + * They are used to map the physical addresses to virtual addresses.
> + */
> +
> +static void *mailbox_rx_base;
> +static void *mailbox_tx_base;
> +static void *mailbox_tx_sram_va;
> +static void *mailbox_rx_sram_va;
> +
> +/*
> + * mailbox_init() - Initialize the mailbox MMIO and SRAM regions
> + */
> +
> +void mailbox_init(void)
>
> Blank line after comments is this a standard ?
> ------------------------------
>
> In core/arch/arm/plat-k3/drivers/mailbox.c
> <https://github.com/AniketSarkar123/optee_os/pull/1#discussion_r2207166906>
> :
>
> > +	}
> +}
> +
> +/*
> + * Function to poll for mailbox rx messages
> + */
> +
> +static int8_t ti_mailbox_poll_rx_status(void)
> +{
> +	uint32_t num_messages_pending = 0U;
> +	uint32_t retry_count = 100U;
> +
> +	while (num_messages_pending == 0U) {
> +		num_messages_pending = io_read32((vaddr_t)mailbox_rx_base +
> +		TI_MAILBOX_MSG_STATUS);
> +
>
> alignment
> ------------------------------
>
> In core/arch/arm/plat-k3/drivers/mailbox.c
> <https://github.com/AniketSarkar123/optee_os/pull/1#discussion_r2207168025>
> :
>
> > + * Function to poll for mailbox rx messages
> + */
> +
> +static int8_t ti_mailbox_poll_rx_status(void)
> +{
> +	uint32_t num_messages_pending = 0U;
> +	uint32_t retry_count = 100U;
> +
> +	while (num_messages_pending == 0U) {
> +		num_messages_pending = io_read32((vaddr_t)mailbox_rx_base +
> +		TI_MAILBOX_MSG_STATUS);
> +
> +		if (retry_count-- == 0U) {
> +			EMSG("Mailbox RX status polling timed out");
> +			return -1;
> +		}
>
> Missed replacing -1
> ------------------------------
>
> In core/arch/arm/plat-k3/drivers/mailbox.c
> <https://github.com/AniketSarkar123/optee_os/pull/1#discussion_r2207169217>
> :
>
> > +	uint32_t num_messages_pending = 0U;
> +	uint32_t retry_count = 100U;
> +
> +	while (num_messages_pending == 0U) {
> +		num_messages_pending = io_read32((vaddr_t)mailbox_rx_base +
> +		TI_MAILBOX_MSG_STATUS);
> +
> +		if (retry_count-- == 0U) {
> +			EMSG("Mailbox RX status polling timed out");
> +			return -1;
> +		}
> +		mdelay(10);
> +	}
> +
> +	return 0;
> +}
>
> Add proper return Macro here
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AniketSarkar123/optee_os/pull/1#pullrequestreview-3019807466>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A4XDCCK7JFIKK2TKDCARTTT3ITOLZAVCNFSM6AAAAACA5VXBWOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMJZHAYDONBWGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

AniketSarkar123 commented on core/arch/arm/plat-k3/drivers/ti_sci_transport.h:
```diff
@@ -0,0 +1,74 @@
+ *
+ * Return: 0 on success, or an error code on failure.
+ */
+TEE_Result ti_sci_clear_init(void);
```
correct the function explanation

AniketSarkar123 commented on core/arch/arm/plat-k3/platform_config.h:
```diff
@@ -54,6 +54,13 @@
 #define SEC_PROXY_REQUEST_THREAD        11
 #endif
 #define OPTEE_HOST_ID                   11
+#define TI_SCI_HOST_ID			10
```
move inside am62l ifdef

AniketSarkar123 replied:
Dont move TI_SCI_HOST_ID that is common

AniketSarkar123 commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,188 @@
+		EMSG("Failed to map mailbox MMIO or SRAM regions");
+		return;
+	}
+}
```
align properly

Pratham-T commented on core/arch/arm/plat-k3/drivers/mailbox.c:
```diff
@@ -0,0 +1,171 @@
+	paddr_t pa;
+
+	if (!msg)
+		return TEE_ERROR_BAD_STATE;
```
Return appropriate error code. For instance, here it should be TEE_ERROR_BAD_PARAMETERS

Pratham-T commented on core/arch/arm/plat-k3/drivers/sec_proxy.c:
```diff
@@ -61,6 +63,7 @@ struct k3_sec_proxy_thread {
  *				  sending/receiving data
  * @dir: Direction of the thread
  */
+
```
Remove extra newline

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -527,7 +528,6 @@ int ti_sci_init(void)
 {
 	struct ti_sci_msg_resp_version rev_info = { };
 	int ret = 0;
-
```
Retain this newline between variable declaration and code.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci_transport.h:
```diff
@@ -0,0 +1,74 @@
+	uint8_t *buf;
+};
+
+/**
+ * mailbox_init() - Initialize the mailbox MMIO and SRAM regions
+ *
+ * This function initializes the mailbox MMIO and SRAM regions used for
+ * sending and receiving messages.
+ */
+void mailbox_init(void);
```
This is a mailbox specific function. Is this needed to be declared here?

If not, remove it.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci_transport.h:
```diff
@@ -0,0 +1,74 @@
+ */
+void mailbox_init(void);
+
+/**
+ * ti_mailbox_poll_rx_status() - Poll the mailbox RX status
+ *
+ * This function polls the mailbox RX status to check if a message has
+ * been received.
+ *
+ * Return: 0 on success, or an error code on failure.
+ */
+TEE_Result ti_mailbox_poll_rx_status(void);
```
Same as `mailbox_init()`

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -153,15 +154,16 @@ int ti_sci_get_revision(struct ti_sci_msg_resp_version *rev_info)
 	struct ti_sci_msg_req_version req = { };
 	struct ti_sci_xfer xfer = { };
 	int ret = 0;
-
```
Retain this newline between variable declaration and code.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -153,15 +154,16 @@ int ti_sci_get_revision(struct ti_sci_msg_resp_version *rev_info)
 				&req, sizeof(req),
 				rev_info, sizeof(*rev_info),
 				&xfer);
+
```
Newline here unnecessary.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -153,15 +154,16 @@ int ti_sci_get_revision(struct ti_sci_msg_resp_version *rev_info)
 		return ret;
 
 	ret = ti_sci_do_xfer(&xfer);
+
```
Same here.

Pratham-T commented on core/arch/arm/plat-k3/drivers/ti_sci.c:
```diff
@@ -313,7 +315,6 @@ int ti_sci_get_dkek(uint8_t sa2ul_instance,
 	struct ti_sci_msg_resp_sa2ul_get_dkek resp = { };
 	struct ti_sci_xfer xfer = { };
 	int ret = 0;
-
```
Retain this newline between variable declaration and code.

Pratham-T reviewed (commented):
Squash this commit ([refactor(ti): abstract sec_proxy and mailbox drivers](https://github.com/AniketSarkar123/optee_os/pull/1/commits/d565a95a5604e14ea5844859b613b9a9f504390f)) into previous ([feat(ti): add support for TI mailbox driver](https://github.com/AniketSarkar123/optee_os/pull/1/commits/a74818c42a06f3c41ca210d7d773b071b67e71fb))

Pratham-T commented:
With these latest comments resolved, the PR overall is in a good shape and LGTM.

AniketSarkar123 committed:
plat-k3: drivers: Add memory mappings for mailbox

Physical memory is required to be mapped to virtual memory
in OP-TEE. All important registers to be used in mailbox are
mapped here.

Signed-off-by: Aniket Sarkar <a-sarkar1@ti.com>

AniketSarkar123 committed:
plat-k3: drivers: Fixing TISCI max message size in sec_proxy

Changing the max message size of TISCI message to correct value.

Signed-off-by: Aniket Sarkar <a-sarkar1@ti.com>
